From c8978b7be6dbe388596fb899ab41a29e414ea5dc Mon Sep 17 00:00:00 2001
From: Daniel Mihai <Daniel.Mihai@microsoft.com>
Date: Wed, 28 Jul 2021 14:55:12 -0700
Subject: [PATCH] Replacing deprecated functions with NULL or highest
 supported.

This is a workaround until OpenSSL issue #7048 is officially resolved.
Issue link: https://github.com/openssl/openssl/issues/7048.

The main purpose of the change is to prevent breaking applications
as they dynamically link to 'libssl.so' where APIs for some
deprecated protocols are no longer present. With this change
OpenSSL's build time configuration may skip the 'no-<prot>-method'
switch, while still not supporting the deprecated protocols disabled
through the 'no-<prot>' switch.

For deprecated DTLS protocol versions behind the scenes we're calling
into 'DTLS_(client_|server_)?method()' set of methods, which
automatically negotiate the highest supported protocol.

For SSLv3 methods we're returning a NULL pointer as there are no
more supported methods for the SSL protocol.
---
 ssl/methods.c | 18 +++++++++++++++---
 1 file changed, 15 insertions(+), 3 deletions(-)

diff --git a/ssl/methods.c b/ssl/methods.c
index c846143277..a7ae074bfd 100644
--- a/ssl/methods.c
+++ b/ssl/methods.c
@@ -215,17 +215,29 @@ const SSL_METHOD *TLSv1_client_method(void)
 # ifndef OPENSSL_NO_SSL3_METHOD
 const SSL_METHOD *SSLv3_method(void)
 {
+#  ifdef OPENSSL_NO_SSL3
+    return NULL;
+#  else
     return sslv3_method();
+#  endif
 }
 
 const SSL_METHOD *SSLv3_server_method(void)
 {
+#  ifdef OPENSSL_NO_SSL3
+    return NULL;
+#  else
     return sslv3_server_method();
+#  endif
 }
 
 const SSL_METHOD *SSLv3_client_method(void)
 {
+#  ifdef OPENSSL_NO_SSL3
+    return NULL;
+#  else
     return sslv3_client_method();
+#  endif
 }
 # endif
 
@@ -249,17 +261,17 @@ const SSL_METHOD *DTLSv1_2_client_method(void)
 # ifndef OPENSSL_NO_DTLS1_METHOD
 const SSL_METHOD *DTLSv1_method(void)
 {
-    return dtlsv1_method();
+    return DTLS_method();
 }
 
 const SSL_METHOD *DTLSv1_server_method(void)
 {
-    return dtlsv1_server_method();
+    return DTLS_server_method();
 }
 
 const SSL_METHOD *DTLSv1_client_method(void)
 {
-    return dtlsv1_client_method();
+    return DTLS_client_method();
 }
 # endif
 
-- 
2.25.1

